File: Broker/receiver.py
import logging
import queue
import random

from paho.mqtt import client as mqtt_client

class Receiver:
    def __init__(self, host, queue:queue.Queue) -> None:  
        self.client = mqtt_client.Client(f"receiver-{random.randint(0,100)}")
        self.message_queue = queue
        self.client.on_connect = self.on_connect
        self.client.on_message = self.on_message
        self.host = host
    
    def __del__(self):
        self.client.disconnect()
    
    def start(self, queue:queue.Queue, topic):
        logging.info(f"Started listening for messages on {topic}")
        self.client.connect(self.host)
        self.client.subscribe(topic)
        self.client.loop_start()

    def on_message(self, client, userdata, msg):
        logging.debug("MESSAGE RECEIVED")
        message = msg.payload.decode("utf-8")
        self.message_queue.put(message)

    def on_connect(self, client, userdata, flags, rc):
        if rc == 0:
            logging.info("Connected to MQTT Broker!")
        else:
            logging.error("Failed to connect, return code %d\n", rc)


    
    def stop(self, topic):
        self.client.loop_stop()
        self.client.unsubscribe(topic)

File: RT802_Kossi_AGBODJA_Jean_MASSIP/Broker_New/receiver/receiver.go
package receiver

import (
	"context"
	"fmt"
	"math/rand"

	mqtt "github.com/eclipse/paho.mqtt.golang"
)

type Receiver struct {
	MessageChan           chan *string
	Client                mqtt.Client
	MessageHandler        func(client mqtt.Client, msg mqtt.Message)
	ConnectHandler        func(client mqtt.Client)
	ConnectionLostHandler func(client mqtt.Client, err error)
	cancelFunc            context.CancelFunc
}

func New() (*Receiver, error) {
	r := Receiver{
		MessageChan: make(chan *string),
	}

	r.ConnectionLostHandler = func(client mqtt.Client, err error) {
		fmt.Printf("Connection to Broker lost : %v\n", err)
		r.cancelFunc()
		close(r.MessageChan)
	}

	r.ConnectHandler = func(client mqtt.Client) {
		fmt.Println("Connected !")
	}

	r.MessageHandler = func(client mqtt.Client, msg mqtt.Message) {
		strMsg := string(msg.Payload())
		fmt.Printf("message received : %s\n", strMsg)
		r.MessageChan <- &strMsg
	}

	return &r, nil
}

func (r *Receiver) Connect() context.Context {
	opts := mqtt.NewClientOptions()
	opts.AddBroker("tcp://127.0.0.1:1883")
	opts.SetClientID(fmt.Sprintf("receiver-%d", rand.Intn(1000)))
	opts.SetDefaultPublishHandler(r.MessageHandler)
	opts.OnConnect = r.ConnectHandler
	opts.OnConnectionLost = r.ConnectionLostHandler
	r.Client = mqtt.NewClient(opts)
	if token := r.Client.Connect(); token.Wait() && token.Error() != nil {
		panic(token.Error())
	}

	ctx, cancel := context.WithCancel(context.Background())
	r.cancelFunc = cancel
	return ctx
}

func (r *Receiver) Subscribe(topic string) {
	token := r.Client.Subscribe(topic, 0, nil)
	token.Wait()
	fmt.Printf("Subscribed to %s \n", topic)
}

func (r *Receiver) Unsubscribe(topic string) {
	r.cancelFunc()
	token := r.Client.Unsubscribe(topic)
	token.Wait()
	fmt.Printf("Unsubscribed from %s \n", topic)
}

func (r *Receiver) Disconnect() {
	r.cancelFunc()
	r.Client.Disconnect(2500)
}
